src/test/java/jenkins/plugins/git/GitToolChooserTest.java

```diff
@@ -0,0 +1,478 @@
+    }
+
+    /*
+    In the event of having an extension which is not applicable to the remote URL provided by the git plugin,
+    the estimator recommends no git implementation
+     */
+    @Test
+    public void testSizeEstimationWithBitbucketAPIs() throws Exception {
+        String remote = "https://bitbucket.com/rishabhBudhouliya/git-plugin.git";
+        sampleRepo.init();
+        store.addCredentials(Domain.global(), createCredential(CredentialsScope.GLOBAL, "github"));
+        store.save();
+        buildAProject(sampleRepo, false);
+        List<TopLevelItem> list = jenkins.jenkins.getItems();
+
+        // Assuming no tool is installed by user and git is present in the machine
+        String gitExe = "git";
+
+        GitToolChooser sizeEstimator = new GitToolChooser(remote, list.get(0), "github", gitExe);
+        assertThat(sizeEstimator.getGitTool(), is("NONE"));
+    }
+
+    /*
+    In the event of having an extension which is applicable to the remote URL but throws an exception due to some
+    reason from the implemented git provider plugin, the estimator handles the exception by silently logging an
+    "INFO" message and returns no recommendation.
+     */
+    @Test
+    public void testSizeEstimationWithException() throws Exception {
+        String remote = "https://bitbucket.com/rishabhBudhouliya/git-plugin.git";
+        sampleRepo.init();
+        store.addCredentials(Domain.global(), createCredential(CredentialsScope.GLOBAL, "github"));
+        store.save();
+        buildAProject(sampleRepo, false);
+        List<TopLevelItem> list = jenkins.jenkins.getItems();
+
+        // Assuming no tool is installed by user and git is present in the machine
+        String gitExe = "git";
+
+        GitToolChooser sizeEstimator = new GitToolChooser(remote, list.get(0), "github", gitExe);
+
+        assertThat(sizeEstimator.getGitTool(), is("NONE"));
+    }
+
+    /*
+    In case of having no user credentials, the git plugin expects the `implementers` of the extension point to handle
+    and try querying for size of repo, if it throws an exception we catch it and recommend "NONE", i.e, no recommendation.
+     */
+    @Test
+    public void testSizeEstimationWithNoCredentials() throws Exception {
+        sampleRepo.init();
+
+        buildAProject(sampleRepo, true);
+        List<TopLevelItem> list = jenkins.jenkins.getItems();
+
+        // Assuming no tool is installed by user and git is present in the machine
+        String gitExe = "git";
+
+        GitToolChooser sizeEstimator = new GitToolChooser(sampleRepo.toString(), list.get(0), null, gitExe);
+
+        assertThat(sizeEstimator.getGitTool(), is("NONE"));
+    }
+
+    /*
+    Tests related to git tool resolution
+    Scenario 1: Size of repo is < 5 MiB, "jgit" should be recommended
+     */
+    @Test
+    public void testGitToolChooserWithLessThan5Mb() throws Exception {
+        String url = "https:github.com";
```

Comment:
I am not able to understand this `url`. In the earlier tests `remote` was directly being used. 

1. How does `url` differ from `remote` when the GitToolChooser is expecting `remoteName`?
2. What is this format of `url` being used, I don't see it as a proper url pattern.

Am I missing something? @rishabhBudhouliya

Reply:
@omkar-dsd Apologies, I forgot to remove these placeholder urls. I have replaced them with the appropriate remote name.

Furthermore, the remote name/urls do not matter much as the validation logic written within the `TestExtension` is pretty naive, it only checks if the remote url contains the word "github".